src/engine/input/input.main.ts

```diff
@@ -1,91 +1,143 @@
 
 /*********
  * Types *
  ********/
 
 export interface InputModuleState {
-  sharedInputState: SharedInputState;
-  inputState: InputState;
+  inputRingBuffer: InputRingBuffer;
 }
 
 /******************
  * Initialization *
  *****************/
 
+// max ringbuffer items
+const MAX_ITEMS = 100;
+
 export const InputModule = defineModule<IMainThreadContext, InputModuleState>({
   name: "input",
   create(ctx, { sendMessage }) {
-    const inputState = createObjectBufferView(inputStateSchema, ArrayBuffer);
-    const sharedInputState = createTripleBufferBackedObjectBufferView(
-      inputStateSchema,
-      inputState,
-      ctx.gameToMainTripleBufferFlags
-    );
+    const inputRingBufferSab = RingBuffer.getStorageForCapacity(MAX_ITEMS, Float32Array);
+    const inputRingBuffer = createInputRingBuffer(new RingBuffer(inputRingBufferSab, Float32Array as any));
 
-    sendMessage<InitializeInputStateMessage>(Thread.Game, InputMessageType.InitializeInputState, { sharedInputState });
+    sendMessage<InitializeInputStateMessage>(Thread.Game, InputMessageType.InitializeInputState, {
+      inputRingBufferSab,
+    });
 
     return {
-      sharedInputState,
-      inputState,
+      inputRingBuffer,
     };
   },
   init(ctx) {
-    const { inputState } = getModule(ctx, InputModule);
+    const { inputRingBuffer } = getModule(ctx, InputModule);
     const { canvas } = ctx;
 
+    const last: { [key: string]: boolean } = {};
+
     function onMouseDown({ buttons }: MouseEvent) {
       if (document.pointerLockElement === canvas) {
-        inputState.mouseButtons[0] = buttons;
+        switch (buttons) {
+          case 1:
+            if (!enqueueInputRingBuffer(inputRingBuffer, KeyCodes.Mouse1, 1)) {
+              console.warn("input ring buffer full");
+            }
+            break;
+          case 2:
+            if (!enqueueInputRingBuffer(inputRingBuffer, KeyCodes.Mouse2, 1)) {
+              console.warn("input ring buffer full");
+            }
+            break;
+          case 4:
+            if (!enqueueInputRingBuffer(inputRingBuffer, KeyCodes.Mouse3, 1)) {
+              console.warn("input ring buffer full");
+            }
+            break;
+          case 8:
+            if (!enqueueInputRingBuffer(inputRingBuffer, KeyCodes.Mouse4, 1)) {
+              console.warn("input ring buffer full");
+            }
+            break;
+          case 16:
+            if (!enqueueInputRingBuffer(inputRingBuffer, KeyCodes.Mouse5, 1)) {
+              console.warn("input ring buffer full");
+            }
+            break;
+        }
       } else {
         canvas.requestPointerLock();
       }
     }
 
     function onMouseUp({ buttons }: MouseEvent) {
       if (document.pointerLockElement === canvas) {
-        inputState.mouseButtons[0] = buttons;
+        switch (buttons) {
+          case 1:
+            if (!enqueueInputRingBuffer(inputRingBuffer, KeyCodes.Mouse1, 0)) {
```

Comment:
`buttons` is actually a bitset so you should only need one enqueue call here https://developer.mozilla.org/en-US/docs/Web/API/MouseEvent/buttons